collab-service/src/tunneller/collab_tunnel_bridge.ts

```diff
@@ -0,0 +1,236 @@
+    let dataToSend = message.data;
+    // Ignore self-echo
+    if (sender === this.username) {
+      return;
+    }
+    // Handle internal redis message cases
+    switch (message.flag) {
+      case ConnectionOpCode.DATA: // Receive normal, Do nothing
+        break;
+
+      case ConnectionOpCode.JOIN: // Receive A, Send B
+        Logger.info(`${this.username} received JOIN from ${sender}`);
+        await this.pubsub.pushMessage(createAckMessage(this.username, this.nickname));
+        break;
+
+      case ConnectionOpCode.ACK: // Receive B, 'Connected'
+        Logger.info(`${this.username} received ACK from ${sender}`);
+        break;
+
+      case ConnectionOpCode.ROOM_DISCOVER: // Receive ARP discovery, Send 'Who Am I'
+        Logger.info(`${this.username} received ARP from ${sender}`);
+        await this.pubsub.pushMessage(createHelloMessage(this.username));
+        break;
+
+      case ConnectionOpCode.ROOM_HELLO: // Receive 'Who Am I', Save attempt
+        Logger.info(`${this.username} received WMI from ${sender}`);
+        // Complete saving snapshot
+        this.attemptCache.setUsers([this.username, sender]);
+        dataToSend = await this.saveAttempt();
+        await this.pubsub.pushMessage(createDataMessage(this.username, dataToSend));
+        break;
+
+      default:
+        Logger.error('Unknown connection flag');
+        return;
+    }
+    this.call.write(makeDataResponse(dataToSend));
+  }
+
+  /**
+   * Handles connection messages from client
+   * @param request
+   */
+  async handleClientMessage(request: CollabTunnelRequest) {
+    // Ignore heartbeat
+    if (isHeartbeat(request.flags)) {
+      return;
+    }
+    // Handle external client message cases
+    switch (readConnectionOpCode(request.data)) {
+      case OPCODE_QUESTION_REQ: // Retrieve question, send question back to user
+        Logger.info(`${this.username} requested for question`);
+        await this.writeQuestionToClient();
+        break;
+
+      case OPCODE_SAVE_CODE_REQ: // Snapshot code
+        Logger.info(`${this.username} requested for saving code`);
+        await this.handleIncomingSaveRequest(request);
+        break;
+
+      default: // Normal data, push to publisher
+        await this.pubsub.pushMessage(createDataMessage(this.username, request.data));
+    }
+  }
+
+  /**
+   * Handles connection messages: save code request, from client
+   * @param request
+   * @private
+   */
+  private async handleIncomingSaveRequest(request: CollabTunnelRequest) {
+    // Prepare saving snapshot
+    const questionResponse = await getQuestionRedis(this.roomId, this.redis);
+    this.attemptCache.setQuestion(questionResponse);
+    this.attemptCache.setLangContent(request.data);
+    this.attemptCache.setUploader(async (msg: HistoryAttempt) => {
+      const res = await this.historyAgent.uploadHistoryAttempt(msg);
+      if (res.errorMessage) {
+        Logger.error(res.errorMessage);
+      }
+      return res;
+    });
```

Comment:
I believe this Lambda can be shifted to saveAttempt because it does not capture any context.